### ISSUE TITLE
AnnotSV Singularity on an HPC Error `command not found`

### ISSUE THREAD

**Issue by yr542** [lgmgeo/AnnotSV#184]
I am trying to use AnnotSV on my Illumina Manta vcf outputs specifically the diploid vcfs. However, I cannot seem to make it work with a HPO in Singularity.

There are 2 input variables:

- `singularity_container`="/path/to/annotsv-3.3.4--py311hdfd78af_0.sif"

- `input_dir`="path/to/Illumina_Manta_outputs

The code I used:

```
# Loop through the files in the input directory
for file in "$input_dir"/*.vcf; do
    # Get the base name of the file without extension
    filename=$(basename "$file" .vcf)

    # Create the output subdirectory within the input directory
    output_dir_parent="$(dirname "$input_dir")"
    decompressed_dir="$output_dir_parent/Step_3___AnnotSV_Main"
    mkdir -p "$decompressed_dir"
    echo "Output directory: $decompressed_dir"
    output_dir="$decompressed_dir/AnnotSV_Batch_Copied"
    mkdir -p "$output_dir"

    # Set the output file path with "_AnnotSV" appended
    output_file="$output_dir/${filename}_AnnotSV_"

    # Run AnnotSV via Singularity with the base name as output prefix
    singularity exec "$singularity_container" /usr/local/etc/AnnotSV/configfile -SVinputFile "$file" -genomeBuild GRCh38 -hpo "HP:0000009" -outputFile "$output_file" 
done
```

Yet I get the error:

```
/usr/local/etc/AnnotSV/configfile: line 108: P_loss_phen: command not found
/usr/local/etc/AnnotSV/configfile: line 109: P_loss_hpo: command not found
/usr/local/etc/AnnotSV/configfile: line 110: P_loss_source: command not found
/usr/local/etc/AnnotSV/configfile: line 111: P_loss_coord: command not found
/usr/local/etc/AnnotSV/configfile: line 112: P_ins_phen: command not found
/usr/local/etc/AnnotSV/configfile: line 113: P_ins_hpo: command not found
/usr/local/etc/AnnotSV/configfile: line 114: P_ins_source: command not found
/usr/local/etc/AnnotSV/configfile: line 115: P_ins_coord: command not found
/usr/local/etc/AnnotSV/configfile: line 116: P_inv_phen: command not found
/usr/local/etc/AnnotSV/configfile: line 117: P_inv_hpo: command not found
/usr/local/etc/AnnotSV/configfile: line 118: P_inv_source: command not found
/usr/local/etc/AnnotSV/configfile: line 119: P_inv_coord: command not found
/usr/local/etc/AnnotSV/configfile: line 120: po_P_gain_phen: command not found
``` 
but when I try to not define a path in the Singularity it tells me that it cannot find a refGene.txt. What do I do for it to get to work for my HPO. My HPO will change based on what is required.

**Comment by tsnowlan** [comment]
Singularity is affected by your local environment, so it can be a bit finicky. Try using `singularity exec --cleanenv ...`, it can solve a lot of mysterious / hard to troubleshoot issues.

Beyond that, it seems like it might be trying to run the tcl file with bash. Can you post the Singularity definition file? It's usually at `/.singularity.d/Singularity` inside the image.

**Comment by lgmgeo** [comment]
Thanks Tor for helping @yr542

**Comment by yr542** [comment]
How would I do that I can only tell you that I downloaded it?

**Comment by yr542** [comment]
`singularity exec --cleanenv "$singularity_container" /usr/local/etc/AnnotSV/configfile -SVinputFile "$file" -genomeBuild GRCh38 -hpo "HP:0001249" -outputFile "$output_file" -annotationsDir "$phenotype_to_genes"` did not change the error that persists ```/usr/local/etc/AnnotSV/configfile: line 9: -annotationMode:: command not found
/usr/local/etc/AnnotSV/configfile: line 11: -benignAF:: command not found
/usr/local/etc/AnnotSV/configfile: line 13: -candidateGenesFiltering:: command not found
/usr/local/etc/AnnotSV/configfile: line 15: -genomeBuild:: command not found
/usr/local/etc/AnnotSV/configfile: line 17: -hpo:: command not found
/usr/local/etc/AnnotSV/configfile: line 19: -includeCI:: command not found``` and the error continues.

**Comment by tsnowlan** [comment]
Ah, I see now. You're calling the configfile, which the image tries to interpret using `/bin/sh`, rather than `AnnotSV`. Try with:

```shell
singularity exec --cleanenv "$singularity_container" AnnotSV \
    -SVinputFile "$file" \
    -genomeBuild GRCh38 \
    -hpo "HP:0001249"
```

**Comment by yr542** [comment]
When I did that I got this error ```...checking the annotation data sources (June 14 2023 - 09:54)
############################################################################
"/usr/local/share/AnnotSV/Annotations_Human/Genes/GRCh38/refGene.txt.gz" file doesn't exist
Please check your install - Exit with error.
############################################################################```

**Comment by lgmgeo** [comment]
Can you try to bind your `annotationsDir` in Singularity?
```
singularity exec --bind /usr/local/share/AnnotSV/Annotations_Human \
    --cleanenv "$singularity_container" AnnotSV \
    -SVinputFile "$file" \
    -genomeBuild GRCh38 \
    -annotationsDir /usr/local/share/AnnotSV/Annotations_Human \
    -hpo "HP:0001249"
```

**Comment by kiranpatil222** [comment]
If your SIF file works fine,, Please let me know the working command and Please provide Singularity DEF file for it

**Comment by yr542** [comment]
I got the singularity file for AnnotSV when I did this `singularity pull docker://quay.io/biocontainers/annotsv` I do not believe I can attach my singularity file here?

**Comment by kiranpatil222** [comment]
Oohh I See.. Biggest issue with Docker / Singularity when we pull,, there isnt any instructions on how to use its executables,, i also have had similar issues when i pull docker imager through singularity but then No idea how to use it

**Comment by yr542** [comment]
Usually it goes with `singularity exec` but there is something that doesn't work well with the refGene this is why I have this post

**Comment by nvnieuwk** [comment]
Hi can you try binding the directory with your input files and annotations directory to the container? This can be done with `--bind` (see [here](https://docs.sylabs.io/guides/3.0/user-guide/bind_paths_and_mounts.html#user-defined-bind-paths) for more information about this).

**Comment by yr542** [comment]
I do not believe this AnnotSV singularity works does anyone have a working example using this Singularity?

**Comment by nvnieuwk** [comment]
I'm using it in a nextflow pipeline on an HPC cluster and it works perfectly for me...

**Comment by yr542** [comment]
Could I have the code you used? As it does not seem to work for me?

**Comment by nvnieuwk** [comment]
Nextflow arranges the execution of singularity containers internally, so I'm not sure it'll help you, but here is the code that does the execution of AnnotSV: https://github.com/CenterForMedicalGeneticsGhent/nf-cmgg-structural/blob/add-expansionhunter/modules/nf-core/annotsv/annotsv/main.nf

**Comment by kiranpatil222** [comment]
Did anyone able to run above annotsv with singularity

**Comment by lgmgeo** [comment]
Sorry, I never evaluated AnnotSV via Singularity ;o(

**Comment by nvnieuwk** [comment]
I'm not sure either why it's not working for everyone else. I didn't create the container myself (it's automatically generated via bioconda). Maybe you could ask more info on the bioconda [gitter](https://app.gitter.im/#/room/#bioconda_Lobby:gitter.im)?

**Comment by yr542** [comment]
I think this means that the AnnotSV does not work with Singularity?

**Comment by nvnieuwk** [comment]
Have you tried converting the docker container to a singularity image? maybe this could work?

**Comment by yr542** [comment]
I can only put a Singularity on the cluster. Is there a way to use the Github repository and then a phenotype to genes file to make HPO terms?
